File: Data Structures & Algorithms/unboundedKnapsack/submission-1.cpp
class Solution {
public:
    int maximumProfit(vector<int>& val, vector<int>& wt, int cap) {
        int n = wt.size();
        vector<vector<int>> dp(n, vector<int>(cap + 1, 0));

        for(int j = 0; j <= cap; j++) {
            if(wt[0] <= j) dp[0][j] = val[0];
        }

        for(int i = 1; i < n; i++) {
            for(int j = 1; j <= cap; j++) {
                int skip = dp[i-1][j];
                int take = 0;

                if(wt[i] <= j) {
                    take = val[i] + dp[i][j - wt[i]];
                }

                dp[i][j] = max(skip, take);
            }
        }

        return dp[n-1][cap];
    }
};

File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def act(i: int, dp: List[int]) -> int:
            if i >= len(nums):
                return 0
            if dp[i] != -1:
                return dp[i]
            skip = 0 + act(i+1, dp)
            take = nums[i] + act(i+2, dp)
            dp[i] = max(skip, take)
            return dp[i]

        dp = [-1] * (len(nums) + 2)
        return act(0, dp)
        